crates/ruff_server/src/session/index.rs

```diff
@@ -159,14 +156,7 @@ impl Index {
                 self.notebook_cells
                     .insert(opened_cell.uri.clone(), path.clone());
             }
-            for closed_cell in did_close.iter().flatten() {
-                if self.notebook_cells.remove(&closed_cell.uri).is_none() {
-                    tracing::warn!(
-                        "Tried to remove a notebook cell that does not exist: {}",
-                        closed_cell.uri
-                    );
-                }
-            }
```

Comment:
I'm not sure if we should remove this logic. Does VS Code send two requests (`notebookDocument/didChange` and `textDocument/didClose`) when a cell is deleted? Is there a case where it might only send either one of them? Can you check which request is sent by VS Code between `notebookDocument/didChange` and `textDocument/didClose` assuming that's what's causing this bug?

Looking at the `pygls` implementation, it retains both logic except that it doesn't warn the user when removing a cell if it doesn't exists: https://github.com/openlawlibrary/pygls/blob/da9b77b7446ffc3c8a503670408dfc76e4a777b0/pygls/workspace/workspace.py#L272-L273

Reply:
Ok, I don't see any additional requests sent by VS Code that references the closed cell between `notebookDocument/didChange` and `textDocument/didClose`:
```
2024-06-14 08:11:46.288 [info] [Trace - 8:11:46 AM] Sending notification 'notebookDocument/didChange'.
2024-06-14 08:11:46.289 [info] Params: {
    "notebookDocument": {
        "version": 7,
        "uri": "untitled:Untitled-1.ipynb?jupyter-notebook"
    },
    "change": {
        "cells": {
            "structure": {
                "array": {
                    "start": 1,
                    "deleteCount": 1
                },
                "didOpen": [],
                "didClose": [
                    {
                        "uri": "vscode-notebook-cell:Untitled-1.ipynb?jupyter-notebook#W2sdW50aXRsZWQ%3D"
                    }
                ]
            }
        }
    }
}


2024-06-14 08:11:46.290 [info]   96.769647s DEBUG ruff_server::session::index Falling back to configuration of the only active workspace for the new document 'untitled:Untitled-1.ipynb?jupyter-notebook'.
  96.769686s DEBUG ruff_server::session::index Falling back to configuration of the only active workspace for the new document 'untitled:Untitled-1.ipynb?jupyter-notebook'.

2024-06-14 08:11:46.290 [info] [Trace - 8:11:46 AM] Received notification 'textDocument/publishDiagnostics'.
2024-06-14 08:11:46.291 [info] Params: {
    "diagnostics": [],
    "uri": "vscode-notebook-cell:Untitled-1.ipynb?jupyter-notebook#W0sdW50aXRsZWQ%3D",
    "version": 7
}


2024-06-14 08:11:46.303 [info] [Trace - 8:11:46 AM] Sending notification 'textDocument/didClose'.
2024-06-14 08:11:46.303 [info] Params: {
    "textDocument": {
        "uri": "vscode-notebook-cell:Untitled-1.ipynb?jupyter-notebook#W2sdW50aXRsZWQ%3D"
    }
}
```

So, I guess it's just that the server is trying to take the snapshot in `textDocument/didClose` handler but the `notebookDocument/didChange` already removed this cell.

Reply:
Do I understand it correctly that this means that VS code always sends an explicit `didClose` notification for all cells that were also closed in the `didChange` notification?

Reply:
Yes, it seems like so. I think it's because cells are modeled as a text document which means that if a cell is removed it triggers both `notebookDocument/didChange` and `textDocument/didClose` request.

Reply:
@dhruvmanila 
> Can you check which request is sent by VS Code between notebookDocument/didChange and textDocument/didClose assuming that's what's causing this bug?

As you saw, there aren't any requests being sent in between. Rather, the `didClose` handler tries to take a document snapshot before removing the document. This is so that it can publish an empty diagnostics list for the document/cell before closing it.